primitives/src/pcs/poly.rs

```diff
@@ -0,0 +1,361 @@
+                    Both(x, y) => x + y,
+                    Left(x) | Right(x) => x,
+                })
+                .collect();
+            Self::from_coeff_vec(coeffs)
+        };
+
+        res.truncate_leading_zeros();
+        res
+    }
+}
+
+impl<T, F> fmt::Debug for GeneralDensePolynomial<T, F>
+where
+    T: GroupCoeff<F>,
+    F: Field,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        for (i, coeff) in self.coeffs.iter().enumerate().filter(|(_, c)| !c.is_zero()) {
+            if i == 0 {
+                write!(f, "\n{:?}", coeff)?;
+            } else if i == 1 {
+                write!(f, " + \n{:?} * x", coeff)?;
+            } else {
+                write!(f, " + \n{:?} * x^{}", coeff, i)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::*;
+    use ark_bls12_381::{Fr, G1Projective};
+    use ark_ec::{short_weierstrass::SWCurveConfig, CurveGroup};
+    use ark_poly::{univariate::DensePolynomial, DenseUVPolynomial, Polynomial};
+    use ark_std::iter::successors;
+    use jf_utils::test_rng;
+
+    // helper function to generate all roots of unity for evaluating polynomial with
+    // `num_coeffs` coeffs.
+    pub(crate) fn get_roots_of_unity<F: FftField>(num_coeffs: usize) -> Vec<F> {
+        let size = if num_coeffs.is_power_of_two() {
+            num_coeffs
+        } else {
+            num_coeffs.checked_next_power_of_two().unwrap()
+        } as u64;
+
+        let group_gen = F::get_root_of_unity(size)
+            .expect("Failed to get roots of unity, maybe wronge domain size");
+        successors(Some(F::from(1u32)), |&prev| Some(prev * group_gen))
+            .take(size as usize)
+            .collect()
+    }
+
+    #[test]
+    fn test_poly_eval_single_point() {
+        let mut rng = test_rng();
+        for _ in 0..5 {
+            let degree = rng.gen_range(5..40);
+            let f = GeneralDensePolynomial::<Fr, Fr>::rand(degree, &mut rng);
+            let g = GeneralDensePolynomial::<G1Projective, Fr>::rand(degree, &mut rng);
+            let point = Fr::rand(&mut rng);
+
+            let f_x = f.evaluate(&point);
+            let expected_f_x: Fr =
+                DensePolynomial::from_coefficients_slice(&f.coeffs).evaluate(&point);
+            assert_eq!(f_x, expected_f_x);
+
+            let g_x = g.evaluate(&point);
+            // g(X) = G0 + G1 * X + G2 * X^2 + G3 * X^3 + ...
+            // turn into an MSM between [G0, G1, G2, ...] and [1, x, x^2, x^3, ...]
+            let scalars: Vec<Fr> = successors(Some(Fr::from(1u32)), |&prev| Some(prev * point))
+                .take((g.degree() + 1) as usize)
+                .collect();
+            let expected_g_x =
+                SWCurveConfig::msm(&CurveGroup::normalize_batch(&g.coeffs), &scalars).unwrap();
+            assert_eq!(g_x, expected_g_x);
+        }
+    }
+
+    #[test]
+    fn test_poly_add() {
+        let mut rng = test_rng();
```

Comment:
add a test that checks whether the addition has truncated the leading zeroes.

Reply:
why would leading zeroes matter from the function perspective? that's like an internal details, right?

Reply:
E.g., you should check the degree is correct (e.g. deg = 1 not 2) after adding a1 + b1X + cX^2 with a2 + b2X + (-c)X^2. This is the invariant property of polynomials, not internal details.